File: 프로그래머스/1/172928. 공원 산책/공원 산책.cpp
#include <string>
#include <vector>
#include <map>
using namespace std;

map<char, pair<int, int>> Cmd = {{'E', {0, 1}}, 
                                {'W', {0, -1}},
                                {'N', {-1, 0}},
                                {'S', {1, 0}}};

vector<int> solution(vector<string> park, vector<string> routes) {
    vector<int> answer;
    
    int curX, curY;
    
    for(int i = 0; i < park.size(); ++i)
    {
        for(int j = 0; j < park[i].length(); ++j)
        {
            if (park[i][j] == 'S')
            {
                curX = j;
                curY = i;
                break;
            }
        }
    }
    
    for(int i = 0; i < routes.size(); ++i)
    {
        auto [dy, dx] = Cmd[routes[i][0]];
        int dis = stoi(routes[i].substr(2, 4));
        
        int tempX = curX;
        int tempY = curY;
        bool flag = true;
        for(int j = 0; j < dis; ++j)
        {
            tempX += dx;
            tempY += dy;
            
            if (tempX < 0 || tempX >= park[0].length() ||  
               tempY < 0 || tempY >= park.size() || park[tempY][tempX] == 'X')
            {
                flag = false;
                break;
            }
        }
        
        if (flag)
        {
            curX = tempX;
            curY = tempY;
        }
    }
    
    answer = {curY, curX};
    return answer;
}

File: 프로그래머스/1/172928. 공원 산책/공원 산책.java
class Solution {
    public int[] solution(String[] park, String[] routes) {
        char[][] map = new char[park.length][park[0].length()];
        int x = -1, y = -1;
        for (int i = 0; i < park.length; i++) {
            for (int j = 0; j < park[i].length(); j++) {
                if (park[i].charAt(j) == 'S') {
                    x = j;
                    y = i;
                }
                map[i][j] = park[i].charAt(j);
            }
        }

        for (String route : routes) {
            String[] tokens = route.split(" ");
            String direction = tokens[0];
            int length = Integer.parseInt(tokens[1]);

            int nx = x;
            int ny = y;
            boolean flag = true;
            for (int i = 0; i < length; i++) {
                switch (direction) {
                    case "N":
                        ny -= 1;
                        break;
                    case "S":
                        ny += 1;
                        break;
                    case "W":
                        nx -= 1;
                        break;
                    case "E":
                        nx += 1;
                        break;
                }

                if (nx < 0 || nx >= map[0].length || ny < 0 || ny >= map.length || map[ny][nx] == 'X') {
                    flag = false;
                    break;
                }
            }
            if (flag) {
                x = nx;
                y = ny;
            }
        }


        return new int[]{y, x};
    }
}